### PR TITLE
Feat/custom closed survey

### LINKED ISSUE #457: [FOR-914] Custom 'Survey Closed' message and CTA
**Is your feature request related to a problem? Please describe.**
When a link survey is completed, respondents are faced with this screen:
grafik (1)

**Describe the solution you'd like**
As a user I'd like to control the headline, subheadline, CTA text and link destination.

**Tasks**

* If  surveyType="link" add a Toggle called "Adjust 'Survey Closed' Message" with description "Change the message visitors see when the survey is closed." in the Response Options card.
* If it's active, display two input fields incl. description. Make sure to use the input field from formbricks/ui:
  * Headline
  * Subheadline

grafik (2)

* Make sure these fields are pre-filled with the default values
* Update LinkSurvey component to display custom messages if set. If a Custom Survey Closed message is set, don't render the Create your own Button.

### How we code at Formbricks 🤓

* Everything is type-safe
* All UI components are in the package `formbricks/ui`
* Run `pnpm dev` to find a demo app to test in-app surveys at `localhost:3002`
* We use **chatGPT** to help refactor code. Use our [Formbricks ✨ megaprompt ✨](https://github.com/formbricks/formbricks/blob/main/megaprompt.md) to create the right context before you write your prompt.

<sub>From [SyncLinear.com](https://synclinear.com) | [FOR-914](https://linear.app/formbricks/issue/FOR-914/custom-survey-closed-message-and-cta)</sub>

### PR DESCRIPTION
**Pull request by Pradumn27**
## What does this PR do?

1. Have added the functionality to add a custom survey close message

Fixes #457

https://github.com/formbricks/formbricks/assets/47187878/dafe440a-53c2-4314-925f-0a49640e022e

Have also fixed the following alignment issue on production

Screenshot 2023-07-03 at 1 12 35 AM

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Chore (refactoring code, technical debt, workflow improvements)
- [ ] Enhancement (small improvements)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change adds a new database migration
- [ ] This change requires a documentation update

## How should this be tested?

- Create a new or edit an existing survey
- Under the Response Options tab click on adjust "survey closed" message
- add your desired heading and subheading
- Publish and mark the survey as completed
- Check survey closed screen by visiting the survey url

## Checklist

- [x] Added a screen recording or screenshots to this PR
- [x] Filled out the "How to test" section in this PR
- [x] Read the [contributing guide](https://github.com/formbricks/formbricks/blob/main/CONTRIBUTING.md)
- [x] Self-reviewed my own code
- [x] Commented on my code in hard-to-understand bits
- [x] Ran `pnpm build`
- [x] Checked for warnings, there are none
- [x] Removed all `console.logs`
- [x] Merged the latest changes from main onto my branch with `git pull origin main`
- [x] My changes don't cause any responsiveness issues
- [ ] Updated the Formbricks Docs if changes were necessary

### CODE REVIEW

File: apps/web/app/environments/[environmentId]/surveys/[surveyId]/edit/ResponseOptionsCard.tsx [hunk 1]
```diff
@@ -36,12 +41,52 @@ export default function ResponseOptionsCard({ localSurvey, setLocalSurvey }: Res
     setLocalSurvey({ ...localSurvey, redirectUrl: link });
   };
 
+  const handleCloseSurveyMessageToggle = () => {
+    if (surveyClosedMessageToggle && localSurvey.surveyClosedMessage) {
```
Comment (jobenjada) [inline apps/web/app/environments/[environmentId]/surveys/[surveyId]/edit/ResponseOptionsCard.tsx:45]: What do you think about this refactoring suggestion by chatGPT

`  setSurveyClosedMessageToggle((prev) => !prev);
  if (surveyClosedMessageToggle && localSurvey.surveyClosedMessage) {
    setLocalSurvey((prev) => ({ ...prev, surveyClosedMessage: null }));
  }
};`

File: apps/web/app/environments/[environmentId]/surveys/[surveyId]/edit/ResponseOptionsCard.tsx [hunk 2]
```diff
@@ -16,32 +16,65 @@ export default function ResponseOptionsCard({ localSurvey, setLocalSurvey }: Res
...
     setLocalSurvey({ ...localSurvey, redirectUrl: link });
   };
 
+  const handleCloseSurveyMessageToggle = () => {
+    setSurveyClosedMessageToggle((prev) => !prev);
+
+    if (surveyClosedMessageToggle && localSurvey.surveyClosedMessage) {
+      setLocalSurvey({ ...localSurvey, surveyClosedMessage: null });
+    }
+  };
```
Comment (Pradumn27) [inline apps/web/app/environments/[environmentId]/surveys/[surveyId]/edit/ResponseOptionsCard.tsx:45]: @jobenjada have taken the suggestion into account and also applied it to the above redirectUrl Toggle

Comment (jobenjada) [reply apps/web/app/environments/[environmentId]/surveys/[surveyId]/edit/ResponseOptionsCard.tsx:45]: ah the comment went out, wasnt sure :) I got a type error when I tried it, will check your solution now and pass over to Matti for a final code review 🙌

Comment (Pradumn27) [reply apps/web/app/environments/[environmentId]/surveys/[surveyId]/edit/ResponseOptionsCard.tsx:45]: Yeah, there was a type error but I resolved it :)

**Comment by jobenjada** [comment]
Hey @Pradumn27 this is only FYI - no need to refactor, just wanted to share what GPT thinks:

**State management -** The state is split across multiple useState calls. This is not inherently bad, but there seems to be some duplication between the localSurvey state and other states like redirectUrl and redirectToggle. It might be easier to reason about the state if it's consolidated into the localSurvey state and its setter function setLocalSurvey.

**Effect on mount -** The useEffect is used to sync the local state with the localSurvey prop. This is only done once when the component mounts because the dependency array is empty. If localSurvey changes after the component has mounted, the local state won't update accordingly. If you want to keep the local state in sync with the localSurvey prop, you should include localSurvey in the dependency array.

**Handling user input -** The handlers for the various user inputs (like handleRedirectUrlChange, handleClosedSurveyMessageChange, handleCheckMark, and handleInputResponse) update the local state and the localSurvey state in separate steps. It would be more efficient to calculate the new state once and then use it to update both states.

Here's a simplified version of the component with these improvements:

```
export default function ResponseOptionsCard({ localSurvey, setLocalSurvey }: ResponseOptionsCardProps) {
  const [open, setOpen] = useState(false);
  
  const updateLocalSurvey = (updatedSurvey: Survey) => {
    setLocalSurvey(updatedSurvey);
  };

  const handleRedirectCheckMark = () => {
    const updatedRedirectUrl = localSurvey.redirectUrl ? null : "";
    updateLocalSurvey({ ...localSurvey, redirectUrl: updatedRedirectUrl });
  };

  const handleRedirectUrlChange = (link: string) => {
    updateLocalSurvey({ ...localSurvey, redirectUrl: link });
  };

  const handleCloseSurveyMessageToggle = () => {
    const updatedSurveyClosedMessage = localSurvey.surveyClosedMessage ? null : {
      heading: "Survey Completed",
      subheading: "This free & open-source survey has been closed",
    };
    updateLocalSurvey({ ...localSurvey, surveyClosedMessage: updatedSurveyClosedMessage });
  };

  const handleClosedSurveyMessageChange = ({
    heading,
    subheading,
  }: {
    heading?: string;
    subheading?: string;
  }) => {
    const message = {
      heading: heading ?? localSurvey.surveyClosedMessage?.heading,
      subheading: subheading ?? localSurvey.surveyClosedMessage?.subheading,
    };
    updateLocalSurvey({ ...localSurvey, surveyClosedMessage: message });
  };

  const handleCheckMark = () => {
    const updatedAutoComplete = localSurvey.autoComplete ? null : 25;
    updateLocalSurvey({ ...localSurvey, autoComplete: updatedAutoComplete });
  };

  const handleInputResponse = (e: any) => {
    let value = parseInt(e.target.value);
    if (value < 1) value = 1;
    updateLocalSurvey({ ...localSurvey, autoComplete: value });
  };

  // Rest of the component...
}

```

**Comment by Pradumn27** [comment]
> Hey @Pradumn27 this is only FYI - no need to refactor, just wanted to share what GPT thinks:
> 
> **State management -** The state is split across multiple useState calls. This is not inherently bad, but there seems to be some duplication between the localSurvey state and other states like redirectUrl and redirectToggle. It might be easier to reason about the state if it's consolidated into the localSurvey state and its setter function setLocalSurvey.
> 
> **Effect on mount -** The useEffect is used to sync the local state with the localSurvey prop. This is only done once when the component mounts because the dependency array is empty. If localSurvey changes after the component has mounted, the local state won't update accordingly. If you want to keep the local state in sync with the localSurvey prop, you should include localSurvey in the dependency array.
> 
> **Handling user input -** The handlers for the various user inputs (like handleRedirectUrlChange, handleClosedSurveyMessageChange, handleCheckMark, and handleInputResponse) update the local state and the localSurvey state in separate steps. It would be more efficient to calculate the new state once and then use it to update both states.
> 
> Here's a simplified version of the component with these improvements:
> 
> ```
> export default function ResponseOptionsCard({ localSurvey, setLocalSurvey }: ResponseOptionsCardProps) {
>   const [open, setOpen] = useState(false);
>   
>   const updateLocalSurvey = (updatedSurvey: Survey) => {
>     setLocalSurvey(updatedSurvey);
>   };
> 
>   const handleRedirectCheckMark = () => {
>     const updatedRedirectUrl = localSurvey.redirectUrl ? null : "";
>     updateLocalSurvey({ ...localSurvey, redirectUrl: updatedRedirectUrl });
>   };
> 
>   const handleRedirectUrlChange = (link: string) => {
>     updateLocalSurvey({ ...localSurvey, redirectUrl: link });
>   };
> 
>   const handleCloseSurveyMessageToggle = () => {
>     const updatedSurveyClosedMessage = localSurvey.surveyClosedMessage ? null : {
>       heading: "Survey Completed",
>       subheading: "This free & open-source survey has been closed",
>     };
>     updateLocalSurvey({ ...localSurvey, surveyClosedMessage: updatedSurveyClosedMessage });
>   };
> 
>   const handleClosedSurveyMessageChange = ({
>     heading,
>     subheading,
>   }: {
>     heading?: string;
>     subheading?: string;
>   }) => {
>     const message = {
>       heading: heading ?? localSurvey.surveyClosedMessage?.heading,
>       subheading: subheading ?? localSurvey.surveyClosedMessage?.subheading,
>     };
>     updateLocalSurvey({ ...localSurvey, surveyClosedMessage: message });
>   };
> 
>   const handleCheckMark = () => {
>     const updatedAutoComplete = localSurvey.autoComplete ? null : 25;
>     updateLocalSurvey({ ...localSurvey, autoComplete: updatedAutoComplete });
>   };
> 
>   const handleInputResponse = (e: any) => {
>     let value = parseInt(e.target.value);
>     if (value < 1) value = 1;
>     updateLocalSurvey({ ...localSurvey, autoComplete: value });
>   };
> 
>   // Rest of the component...
> }
> ```

Thanks for sharing this @jobenjada , I had a few viewpoints why this suggestion is not relevant here: 

1. The states being stored are specific to the component and can't be consolidated to the local survey as it is another separate entity and is used in several other places.
2. We only need to sync the state with the localSurvey on mount as post mounting of the component it is synced only via the local states of this component and after the update api call we refresh the page.

**Review by mattinannt — APPROVED** [review]
@Pradumn27 thanks a lot for this feature PR 💪 Everything works great. I just solved the merge conflicts and will now merge the changes into main 😊🎉

**Pull request merged**